Rebasing code to support data stacking and sprites

GridRL opened the pull request:
Hi Peter!

This is a partially improved version of the stacking prototype I made few months ago.

Just like now, only username metadata is mandatory. Stacking happens on a parameter equal to envID if set, otherwise it's randomized, so it can still display agents without a declared envID.

Custom sprites are supported setting optional metadata sprite_id=0.

Automatic variable speed is supported and scales with packets sizes on a log2 scale. Default speed matches the current, which is kinda x8 from game. I added high hard-limits for a single batch of coordinates at 2048 and 10 packets. Everything above that limit will be discarded. Mostly for safety, without abuse or ultra fast environment, I don't think this is needed. I also maintained legacy low speeds for very small packets, mostly for debug.

Visual artifacts of warps are mitigated with two editing, disabling the sprite if coordinates delta is above 1 and Function clipping inside coordConversionFunc using map bounds.

Direction infering is still using global coordinates over local, but that's fixable later if necessary, also considering the feature is off by default.

Extra Hotkeys are added for:

- [W] Wipe all agents
- [S] Toggle the stacking mechanism (default on - stacking active, otherwise always randomize stackID to get a result comparable to the current version)
- [D] Toggle directional sprites (default off - PIXI.TilingSprite is more GPU heavy and kinda niche)

I hope it's all good!

PWhiddy commented:
Hey! I pulled down the code to run locally. when I press d to enable directional sprites I'm not seeing the directions change, am I doing something wrong maybe?

PWhiddy commented on visualizer_live.js:
```diff
@@ -279,79 +347,166 @@ PIXI.Assets.load([
+    let texturesChars = [];
+    let texturesCharsDirectional = [];
+    
+    for(let i = 0; i < 73; ++i){
```
73 is the number of sprites in the spritesheet?

PWhiddy commented on visualizer_live.js:
```diff
@@ -47,15 +49,64 @@ function smoothstep(min, max, value) {
     return x*x*(3 - 2*x);
 }
 
+const createLock = () => {
```
I'm a little confused why a lock is implemented/used? JS is single threaded so all operations are going to be atomic by default.

GridRL commented:
All god on your end, when at the end I added buttons for directional sprites, I introduced a pythonic typo and used self over this, so the comparison of an undefined field will always be false! :p
Last commit should solve this, I also added a check to see if tilePosition is undefined to avoid runtime errors.
Yes, 73 is tied to the amount of sprites in the image. For the lock you are right, it's pretty useless on JS. It was mostly a comment, so it's safe to remove.

PWhiddy commented:
Would be very cool to have this! with the current performance and the extra code though it's not ready for merge yet.
